overlord/servicestate/quota_control.go

```diff
@@ -0,0 +1,232 @@
+func ensureSnapServicesForGroup(st *state.State, grp *quota.Group) error {
+	// build the map of snap infos to options to provide to EnsureSnapServices
+	snapSvcMap := map[*snap.Info]*wrappers.SnapServiceOptions{}
+	for _, sn := range grp.Snaps {
+		info, err := snapstate.CurrentInfo(st, sn)
+		if err != nil {
+			return err
+		}
+
+		opts, err := SnapServiceOptions(st, sn)
+		if err != nil {
+			return err
+		}
+
+		snapSvcMap[info] = opts
+	}
+
+	ensureOpts := &wrappers.EnsureSnapServicesOptions{
+		Preseeding: snapdenv.Preseeding(),
+	}
+
+	// set RequireMountedSnapdSnap if we are on UC18+ only
+	deviceCtx, err := snapstate.DeviceCtx(st, nil, nil)
+	if err != nil {
+		return err
+	}
+
+	if !deviceCtx.Classic() && deviceCtx.Model().Base() != "" {
+		ensureOpts.RequireMountedSnapdSnap = true
+	}
+
+	// TODO: do we need to restart modified services ?
+	return wrappers.EnsureSnapServices(snapSvcMap, ensureOpts, nil, progress.Null)
+}
+
+// RemoveQuota deletes the specific quota group.
+// TODO: currently this only supports removing leaf sub-group groups, it doesn't
+// support removing parent quotas, but probably it makes sense to allow that too
+func (mgr *ServiceManager) RemoveQuota(name string) error {
+	st := mgr.state
+	st.Lock()
+	defer st.Unlock()
+
+	// first get the group for later before it is deleted from state
+	grp, err := GetQuota(st, name)
+	if err != nil {
+		return err
+	}
+
+	// XXX: remove this limitation eventually
+	if len(grp.SubGroups) != 0 {
+		return fmt.Errorf("cannot remove quota group with sub-groups, remove the sub-groups first")
+	}
+
+	// delete the group from state - do this firstfor convenience to ensure that
+	// we can just use SnapServiceOptions below and since it operates via state,
+	// it will immediately reflect the deletion
+	var allGrps map[string]*quota.Group
+
+	if err := st.Get("quotas", &allGrps); err != nil {
+		return err
+	}
+
+	delete(allGrps, name)
+	st.Set("quotas", allGrps)
+
+	if err := ensureSnapServicesForGroup(st, grp); err != nil {
+		return err
+	}
+
+	// seperately delete the slice unit, EnsureSnapServices does not do this for
+	// us
+	if err := wrappers.RemoveQuotaGroup(grp); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (mgr *ServiceManager) AddSnapToQuota(grpName string, snapName string) error {
+	st := mgr.state
+	st.Lock()
+	defer st.Unlock()
+
+	grp, err := GetQuota(st, grpName)
+	if err != nil {
+		return err
+	}
+
+	// check if the snap is already in the group
+	for _, subGrp := range grp.Snaps {
+		if subGrp == snapName {
+			// we are done, it's already in the group
+			return nil
+		}
+	}
+
+	// add the snap and save it to state
+	grp.Snaps = append(grp.Snaps, snapName)
+	if err := UpdateQuotas(st, grp); err != nil {
+		return err
+	}
+
+	// update the snap services
+	if err := ensureSnapServicesForGroup(st, grp); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (mgr *ServiceManager) ChangeGroupMemoryLimit(grpName string, memoryLimit quantity.Size) error {
```

Comment:
This is related to the previous question and the general one I asked under the PR; I think we should have a way to update all aspects of the group at once.

Reply:
this has been refactored to use a more generic UpdateQuota method which hopefully is as generic as we need it to be